design/replace-CLI-with-SDK.md

```diff
@@ -0,0 +1,220 @@
+	s.options.DoNotVerifyTLS, _ = utils.GetBoolOrDefault(options[kopialib.DoNotVerifyTLS], true)
+}
+
+```
+
+
+#### Repository pkg
+
+This pkg would be a wrapper over the `storage pkg` built above and the kopia repositoy pkg `github.com/kopia/kopia/repo` 
+provided by kopia SDK
+
+```go
+
+package repository
+
+type Repository struct {
+	st          storage.Storage
+	password    string
+	configFile  string
+	storageType storage.StorageType
+}
+
+// Create repository using kopia SDK
+func (r *Repository) Create(opts *repo.NewRepositoryOptions) (err error) {
+	storage, err := r.st.Connect()
+	if err != nil {
+		return err
+	}
+	return repo.Initialize(context.Background(), storage, opts, r.password)
+}
+
+// Connect to the repository using kopia SDK
+func (r *Repository) Connect(opts *repo.ConnectOptions) (err error) {
+	storage, err := r.st.Connect()
+	if err != nil {
+		return err
+	}
+	return repo.Connect(context.Background(), r.configFile, storage, r.password, opts)
+}
+
+// Connect to the repository by providing a config file
+func (r *Repository) ConnectUsingFile() error {
+	repoConfig := repositoryConfigFileName(r.configFile)
+	if _, err := os.Stat(repoConfig); os.IsNotExist(err) {
+		return errors.New("failed find kopia configuration file")
+	}
+
+	_, err := repo.Open(context.Background(), repoConfig, r.password, &repo.Options{})
+	return err
+}
+
+
+func repositoryConfigFileName(configFile string) string {
+	if configFile != "" {
+		return configFile
+	}
+	return filepath.Join(os.Getenv("HOME"), ".config", "kopia", "repository.config")
+}
+
+```
+
+### Repository server controller changes
+
+#### Kopia CLI Approach
+
+![Alt text](images/kopia-CLI-workflow.png?raw=true "Kanister Kopia Integration using Kopia CLI")
+
+Above diagram explains the current workflow repository server controller
+uses to start the kopia repository server. All the commands are executed from the controller pod inside 
+the respoitory server pod using `kube.exec`. The repository server pod that is created by controller
+uses `kanister-tools` image
+
+
+#### Kopia SDK Approach
+
+![Alt text](images/kopia-SDK-workflow.png?raw=true "Kanister Kopia Integration using Kopia SDK")
+
+
+As shown in the figure we will be building a custom image which is going have this workflow:
+1. Start the kopia repository server in `--async-repo-connect` mode that means the server would be started without
+connecting to the repository in an async mode. Existing approach starts the server only after the connection to kopia repository
+is successful. Kopia SDK currently does not have an exported function to start the kopia server. So we would still be using Kopia
+CLI to start the server
+2. Check kopia server status using Kopia SDK wrappers explained in section [Kopia SDK wrappers](#kopia-sdk-wrappers)
+2. Connect to Kopia repository using Kopia SDK wrappers
+3. Add or Update server users using Kopia SDK wrappers
+4. Refresh server using Kopia SDK wrappers
```

Comment:
Are all the operations performed as a startup process? Or we are exposing APIs that the client will be calling to perform those operations? Or is it a mix of both?

Reply:
All the operations would be part of a startup process